### DISCUSSION TITLE
Why is `LANG` environment not defined when starting MacOS application bundle

### DISCUSSION THREAD

**Discussion by ikus060** [pyinstaller/pyinstaller#6378] [PyInstaller] [answered]
Hello,

Using a MacOs application bundle created using Pyinstaller. When starting the application bundle. The environment variable `LANG` is not defined which causes issue to load our translation for the appropriate language.

Here the difference
```
open /Applications/Minarca.app/

environ({'USER': 'vmtest', 'PATH': '/usr/bin:/bin:/usr/sbin:/sbin', 'LOGNAME': 'vmtest', 'SSH_AUTH_SOCK': '/private/tmp/com.apple.launchd.Gmu9KYb5hJ/Listeners', 'HOME': '/Users/vmtest', 'SHELL': '/bin/zsh', '__CF_USER_TEXT_ENCODING': '0x1F5:0x0:0xB', 'TMPDIR': '/var/folders/38/qfyr8rgx7x17trffq51ghz600000gn/T/', 'XPC_SERVICE_NAME': 'com.ikus-soft.minarca.3884', 'XPC_FLAGS': '0x0', 'TCL_LIBRARY': '/Applications/Minarca.app/Contents/MacOS/tcl', 'TK_LIBRARY': '/Applications/Minarca.app/Contents/MacOS/tk'})
```
VS
```
/Applications/Minarca.app/Contents/MacOS/minarcaw         

environ({'TMPDIR': '/var/folders/38/qfyr8rgx7x17trffq51ghz600000gn/T/', 'XPC_FLAGS': '0x0', 'TERM_PROGRAM_VERSION': '433', 'LANG': 'fr_CA.UTF-8', 'TERM_PROGRAM': 'Apple_Terminal', 'XPC_SERVICE_NAME': '0', 'TERM_SESSION_ID': '40C36F3B-D023-4455-B348-B19F02414D28', 'TERM': 'xterm-256color', 'SSH_AUTH_SOCK': '/private/tmp/com.apple.launchd.Gmu9KYb5hJ/Listeners', 'SHELL': '/bin/zsh', 'HOME': '/Users/vmtest', 'LOGNAME': 'vmtest', 'USER': 'vmtest', 'PATH': '/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin', 'SHLVL': '1', 'PWD': '/Users/vmtest', 'OLDPWD': '/Users/vmtest', '_': '/Applications/Minarca.app/Contents/MacOS/minarcaw', '__CF_USER_TEXT_ENCODING': '0x1F5:0x0:0xB', 'TCL_LIBRARY': '/Applications/Minarca.app/Contents/MacOS/tcl', 'TK_LIBRARY': '/Applications/Minarca.app/Contents/MacOS/tk'})
```

**Comment by rokm** [comment, marked as answer]
>  Why is `LANG` environment not defined when starting MacOS application bundle

Your guess is as good as ours (i.e., you should be asking in Apple forums, not here). 

If I had to take a guess, the app bundles are not considered POSIX applications by macOS, and hence do not receive full POSIX-like environment when launched via Finder. Or perhaps the environment is set up by the shell, and running via `open` bypasses the shell, resulting in limited environment.

**Reply by ikus060** [reply]
I'm asking here mostly because this "application bundle" created by pyinstaller is then broken. `LANG` and other `LC_*` variables are quite important for many application in python and thus, if the application bundle generated by pyinstaller are not receive those variable, the application might be broken.

I'm not the first asking for similar question, but no definitive answer was provided about hos to get it fixed in pyinstaller.

**Reply by bwoodsend** [reply]
Any macOS application bundle (i.e. a .app folder) - be it built by PyInstaller, by hand, or some other application - is run in this limited environment. You won't find a PyInstaller specific solution because it's not a PyInstaller specific problem.

**Reply by ikus060** [reply]
I see.

What's the solution to retrieve the user LANG if the environment variable is not pass to the application bundle then ?

**Reply by rokm** [reply]
> I'm not the first asking for similar question, but no definitive answer was provided about hos to get it fixed in pyinstaller.

Because this is not PyInstaller-specific problem, but rather the way things are with application bundles on macOS. 

If the OS doesn't see it fit to provide a full shell-like environment for application bundles and your application depends on having that, there's nothing we can do about that. And you likely cannot do anything about the former, either, but can adapt the latter to be compatible with the OS you are targeting (along with its idiosyncrasies). I.e., look up the ways to query locale settings on macOS.

**Reply by rokm** [reply]
> look up the ways to query locale settings on macOS.

One way for this:
```python
import Foundation  # pyobjc

sud = Foundation.NSUserDefaults.standardUserDefaults()
locale = sud.stringForKey_('AppleLocale')
```

Not sure if it maps 1:1 to the POSIX locales, though.

**Reply by ikus060** [reply]
Thanks @rokm, That exactly what I need.

Any change PyInstaller use this peace of code to set LANG before start the app ? That would solve a couple of tickets on github.

**Reply by rokm** [reply]
> Any change PyInstaller use this peace of code to set LANG before start the app ? That would solve a couple of tickets on github.

No, because that would require us to depend on `pyobjc`, which might not be desirable in general.
